source/contentRecog/__init__.py

```diff
@@ -0,0 +1,259 @@
+		"""Constructor.
+		@param data: The lines/words data structure. For example:
+			[
+				[
+					{"x": 106, "y": 91, "width": 11, "height": 9, "text": "Word1"},
+					{"x": 117, "y": 91, "width": 11, "height": 9, "text": "Word2"}
+				],
+				[
+					{"x": 106, "y": 105, "width": 11, "height": 9, "text": "Word3"},
+					{"x": 117, "y": 105, "width": 11, "height": 9, "text": "Word4"}
+				]
+			]
+		@type data: list of lists of dicts
+		@param coordConverter: The converter to convert coordinates
+			in the supplied image to screen coordinates.
+			This should be used when returning coordinates to NVDA.
+		@type coordConverter: L{ResultCoordConverter}
+		"""
+		self.data = data
+		self.coordConverter = coordConverter
+		self._textList = []
+		self.textLen = 0
+		#: End offsets for each line.
+		self.lines = []
+		#: Start offsets and screen coordinates for each word.
+		self.words = []
+		self._parseData()
+		self.text = "".join(self._textList)
+
+	def _parseData(self):
+		for line in self.data:
+			firstWordOfLine = True
+			for word in line:
+				if firstWordOfLine:
+					firstWordOfLine = False
+				else:
+					# Separate with a space.
+					self._textList.append(" ")
+					self.textLen += 1
+				self.words.append(LwrWord(self.textLen,
+					self.coordConverter.convertX(word["x"]),
+					self.coordConverter.convertY(word["y"])))
+				text = word["text"]
+				self._textList.append(text)
+				self.textLen += len(text)
+			# End with new line.
+			self._textList.append("\n")
+			self.textLen += 1
+			self.lines.append(self.textLen)
+
+	def makeTextInfo(self, obj, position):
+		return LwrTextInfo(obj, position, self)
+
+class LwrTextInfo(textInfos.offsets.OffsetsTextInfo):
+	"""TextInfo used by L{LinesWordsResult}.
+	This should only be instantiated by L{LinesWordsResult}.
+	"""
+
+	def __init__(self, obj, position, result):
+		self.result = result
+		super(LwrTextInfo, self).__init__(obj, position)
+
+	def copy(self):
+		return self.__class__(self.obj, self.bookmark, self.result)
+
+	def _getTextRange(self, start, end):
+		return self.result.text[start:end]
+
+	def _getStoryLength(self):
+		return self.result.textLen
+
+	def _getLineOffsets(self, offset):
+		start = 0
+		for end in self.result.lines:
+			if end > offset:
+				return (start, end)
+			start = end
+		# offset is too big. Fail gracefully by returning the last line.
+		return (start, self.result.textLen)
+
+	def _getWordOffsets(self, offset):
+		start = 0
+		for word in self.result.words:
+			if word.offset > offset:
+				return (start, word.offset)
+			start = word.offset
+		# offset is in the last word (or offset is too big).
+		return (start, self.result.textLen)
+
+	def _getPointFromOffset(self, offset):
+		word = None
+		for nextWord in self.result.words:
+			if nextWord.offset > offset:
+				# Stop! We need the word before this.
+				break
+			word = nextWord
+		return textInfos.Point(word.left, word.top)
+
+class SimpleTextResult(RecognitionResult):

```

Comment:
This might become clear later in the review and probably needs to be handled higher up than here, but at the very least we would know the area of the screen that the image  used for the OCR came from (assuming nvda did the  screen grab). Obviously this is not true for an image obtained from elsewhere.